pkg/billing/client.go

```diff
@@ -35,16 +35,135 @@ func GetHostname() string {
 	return hostname
 }
 
-func NewClient(url string, c *http.Client) Client {
-	return Client{URL: fmt.Sprintf("%s/usage_events", url), httpc: c}
+type Client interface {
+	Send(ctx context.Context, payload []byte, traceID TraceID) error
+	LogFields() zap.Field
 }
 
 type TraceID string
 
-func (c Client) GenerateTraceID() TraceID {
+func GenerateTraceID() TraceID {
 	return TraceID(shortuuid.New())
 }
 
+type HTTPClient struct {
+	URL   string
+	httpc *http.Client
+}
+
+func NewHTTPClient(url string, c *http.Client) HTTPClient {
+	return HTTPClient{URL: fmt.Sprintf("%s/usage_events", url), httpc: c}
+}
+
+func (c HTTPClient) Send(ctx context.Context, payload []byte, traceID TraceID) error {
+	r, err := http.NewRequestWithContext(ctx, http.MethodPost, c.URL, bytes.NewReader(payload))
+	if err != nil {
+		return RequestError{Err: err}
+	}
+	r.Header.Set("content-type", "application/json")
+	r.Header.Set("x-trace-id", string(traceID))
+
+	resp, err := c.httpc.Do(r)
+	if err != nil {
+		return RequestError{Err: err}
+	}
+	defer resp.Body.Close()
+
+	// theoretically if wanted/needed, we should use an http handler that
+	// does the retrying, to avoid writing that logic here.
+	if resp.StatusCode != http.StatusOK {
+		return UnexpectedStatusCodeError{StatusCode: resp.StatusCode}
+	}
+
+	return nil
+}
+
+func (c HTTPClient) LogFields() zap.Field {
+	return zap.String("url", c.URL)
+}
+
+type S3Client struct {
+	bucket, prefixInBucket string
+	client                 *s3.Client
+	now                    func() time.Time
+}
+
+func NewS3Client(
+	bucket, region, prefixInBucket, endpoint string,
```

Comment:
style nit: separate args should have separate types — IMO this is an anti-pattern that makes the code harder to read and changes more likely to interfere with each other.

Reply:
At the same time:

This looks like a symptom of the underlying issue that there's a bunch of string args in a row with little protection against providing them in the wrong order.

Maybe `pkg/billing` should expose a `S3ClientConfig` that contains these as fields? It could even be directly embedded in `pkg/agent/billing.S3ClientConfig` and passed in, so the args don't need to be rewritten in multiple places.